src/cracking.py

```diff
@@ -0,0 +1,150 @@
+# Debugging
+
+from loguru import logger
+
+# Other
+
+from collections import namedtuple
+from concurrent.futures import ThreadPoolExecutor
+
+# Files
+
+from printing import Prettifier
+import online
+import offline
+
+
+class Searcher:
+    def __init__(self, config):
+        # logger.debug("called class searcher")
+        self.config = config
+        self.searchers_offline = [offline.hashcat()]
+        self.searchers_online = [
+            online.md5crypt(),
+            online.rainbow_tabels(),
+            online.nitrxgen(),
+            online.cmd5(),
+            online.LmRainbowTabels(),
+            online.md5_addr(),
+            online.md5_grom(),
+            online.sha1_grom(),
+            online.hashsorg(),
+        ]
+        self.Hash_input = namedtuple(
+            "Hash_input",
+            ["text", "types", "hashcats", "api_keys", "timeout", "greppable"],
+        )
+
+    def main(self):
+
+        # logger.debug("called main self")
+        return self.perform_search(self.config)
+
+    def perform_search(self, config):
+
+        out = []
+
+        if not config["timeout"]:
+            # logger.debug("Defulting timeout to 1")
+            config["timeout"] = 1
+
+        # print(config["timeout"])
+
+        for hash, types in config["hashes"].items():
+            hash_ctext = hash
+
+            keys = [type["name"].lower() for type in types]
+            hashcats = [type["hashcat"] for type in types]
+
+            # logger.debug(f"Possible types are {keys}")
+
+            # logger.debug(f"Called hash cracker {hash_ctext}")
+
+            supported_searchers = []
+            types = []
+
+            if not config["offline"]:
+                for search in self.searchers_online:
+                    for hashtype in keys:
+                        if hashtype.lower() in search.supports:
+                            if not search in supported_searchers:
+                                supported_searchers.append(search)
+                            if not hashtype in types:
+                                types.append(hashtype)
+
+            for search in self.searchers_offline:

```

Comment:
why do we have this and config["offline"]? Isnt all offline crackers just hashcat / john? DO we need this code?

Reply:
We check for config["offline"] because if its set to True we need to skip over the online searchers